modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTransactionalCache.java

```diff
@@ -627,6 +652,166 @@ else if (!F.isEmpty(req.keys()))
+
+                                map.put(primary, req = new GridNearUnlockRequest(ctx.cacheId(), keyCnt,
+                                    ctx.deploymentEnabled(), fut.futureId()));
+
+                                req.version(ver);
+                                req.forSavepoint();
+
+                                ctx.mvcc().addFuture(fut, fut.futureId());
+
+                                unlockFuts.put(req, fut);
+                            }
+                        }
+
+                        // Remove candidate from local node first.
+                        GridCacheMvccCandidate rmv = entry.removeLock();
+
+                        if (rmv != null) {
+                            if (!rmv.reentry()) {
+                                if (ver != null && !ver.equals(rmv.version()))
+                                    throw new IgniteCheckedException("Failed to unlock " +
+                                        "(if keys were locked separately, " +
+                                        "then they need to be unlocked separately): " + keys);
+
+                                if (!primary.isLocal()) {
+                                    req.addKey(entry.key(), ctx);
+
+                                    unlockFuts.get(req).addKey(key);
+                                }
+                                else
+                                    locKeys.add(cacheKey);
+
+                                if (log.isDebugEnabled())
+                                    log.debug("Removed lock (will distribute): " + rmv);
+                            }
+                            else if (log.isDebugEnabled())
+                                log.debug("Current thread still owns lock (or there are no other nodes)" +
+                                    " [lock=" + rmv + ", curThreadId=" + Thread.currentThread().getId() + ']');
+                        }
+                    }
+
+                    assert !topVer.equals(AffinityTopologyVersion.NONE) || cand == null;
+
+                    if (topVer.equals(AffinityTopologyVersion.NONE))
+                        topVer = ctx.affinity().affinityTopologyVersion();
+
+                    ctx.evicts().touch(entry, topVer);
+
+                    break;
+                }
+                catch (GridCacheEntryRemovedException ignore) {
+                    if (log.isDebugEnabled())
+                        log.debug("Attempted to unlock removed entry (will retry): " + entry);
+                }
+            }
+        }
+
+        if (ver == null)
+            return;
+
+        for (Map.Entry<ClusterNode, GridNearUnlockRequest> mapping : map.entrySet()) {
+            ClusterNode n = mapping.getKey();
+
+            GridDistributedUnlockRequest req = mapping.getValue();
+
+            if (n.isLocal())
+                dht.removeLocks(ctx.nodeId(), req.version(), locKeys, true);
+            else if (!F.isEmpty(req.keys())) {
+                ctx.io().send(n, req, ctx.ioPolicy());
```

Comment:
before sending requests , you must install LockTimeoutObject , like this : cctx.time().addTimeoutObject(timeoutObj);
After receiving all responses you must remove it : cctx.time().removeTimeoutObject(timeoutObj)
So , in case of waiting too long for lock response you will have timeout exception

Reply:
Currently future timeout == transaction timeout, so no need to set it here.

Reply:
the point is you should create the only timeout object instead of many